workspaces/adventure-pack/goodies/typescript/Map.groupBy/index.ts

```diff
@@ -0,0 +1,29 @@
+declare global {
+  interface MapConstructor {
+    groupBy<K, V>(
+      iterable: Iterable<V>,
+      callbackFn: (value: V, index: number) => K,
+    ): Map<K, V[]>;
+  }
+}
+
+Map.groupBy = function <K, V>(
+  iterable: Iterable<V>,
+  callbackFn: (value: V, index: number) => K,
+): Map<K, V[]> {
+  const groups = new Map<K, V[]>();
+  let index = 0;
+  for (const value of iterable) {
+    const key = callbackFn(value, index);
+    if (!groups.has(key)) {
+      groups.set(key, []);
+    }
+    groups.get(key)!.push(value);
+    index++;
```

Comment:
When we don't care about the value before the increment, I advocate for:

```suggestion
    ++index;
```

😇

Reply:
I'd say it's 99+% a style preference 😄 I favor `++foo` in my own code, but I don't think I've ever asked someone to change to it in a code review I've done as part of my paid work. Here, I am mostly allowing myself this indulgence, and exercising my privilege as the person who started the repository 😄 

There is an underlying principle for the preference, however, and whether or not you agree with this being a good application of it or not, I think the principle itself is a good one. Specifically, the principle is an extreme attention to detail and an obsession with how the code style communicates intention or nuance.

For example, if someone starts a string using backticks, like `` `hello ... ``, that communicates the intent "I want to use template strings". My brain gets ready for a template string, and I might read the rest of the string more attentively, preparing for the upcoming `${}`. As such if the string ends up being simply `` `hello world` `` as a reader I would be slightly surprised. I might perhaps even experience a microsecond of disappointment that I turned on my internal`${}` scanner for nothing. The code is in no way incorrect, and I'm not aware of any perf implications, but I'd advocate for using normal quotes in that case, citing the principle of least surprise: why enclose the strings in backticks if you don't need them?

Likewise, when someone writes `foo++`, that might communicate "I would like to use the value before the increment", such that if the value before the increment is never used, it's a minuscule, momentary surprise for a reader. This is a much weaker example than the backticks one, however. Whereas the difference between `foo++` and `++foo` might still be top of mind when writing low-level C or C++ code, I think in JavaScript and other high-level languages that inherited the `++` operators, it's not something that's commonly thought about anymore. I think the zeitgeist is that `foo++` and `++foo` are essentially interchangeable in most cases, and where they're not interchangeable, is the author perhaps being unnecessarily clever?[^1] In this context, I think `foo++` ends up being the natural default, and perhaps its use isn't meant to communicate anything special. Someone could make the argument that `++foo` is inherently more surprising due to being seen less frequently. Maybe that's why languages like Ruby and Python opted to not include the `++` operators at all and avoid any confusion or bikeshedding altogether. Personally, I consider the difference between `foo++` and `++foo` to be within the range of things developers should know. I don't hold that view regarding every operator[^2], but I do hold it here 🤷 

As far as performance implications, I'd expect no difference in any modern runtime. One could imagine the smallest of possible differences in perf in a naive runtime, since `foo++` might need to "store" the old value somehow. But we should be able to determine while parsing the code that the old value isn't needed. And if there _is_ a perf difference, it's probably microscopic compared to the perf hit from using JavaScript in the first place 😄 

Thanks for the opportunity to share my thoughts on this essential matter!

[^1]: I personally try not to rely on the difference between `foo++` and `++foo` in most cases, with a notable exception being `bar[foo++]` which I consider to be a nice idiom for processing a value and advancing a cursor, and one worth being comfortable with.
[^2]: I wouldn't be sad if we lost [the comma operator](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Comma_operator).

Reply:
This was a very clear and thorough explanation! Thank you! 

You put into words

Reply:
This was a very clear and thorough explanation! Thank you!

No worries about exercising your BDFL privilege, I'm now better for it 😄 

I liked what you said about code style and it's relation to the expectations you form while reading it and how it directs your attention. 

Like I said I'm happy to conform here but I'm glad I asked.